packages/components/src/components/SquigglePlayground/index.tsx

```diff
@@ -32,30 +32,40 @@ import {
 } from "../SquiggleViewer/index.js";
 import { ViewSettingsForm, viewSettingsSchema } from "../ViewSettingsForm.js";
 
+import { SqProject } from "@quri/squiggle-lang";
 import { ImportSettingsForm } from "./ImportSettingsForm.js";
 import { RunControls } from "./RunControls/index.js";
 import { useRunnerState } from "./RunControls/useRunnerState.js";
-import { ShareButton } from "./ShareButton.js";
 import {
   EnvironmentSettingsForm,
   playgroundSettingsSchema,
   type PlaygroundFormFields,
 } from "./playgroundSettings.js";
 
-type PlaygroundProps = SquiggleArgs &
-  Omit<SquiggleViewerProps, "result"> & {
-    /** The initial squiggle string to put in the playground */
-    defaultCode?: string;
-    onCodeChange?(expr: string): void;
-    /* When settings change */
-    onSettingsChange?(settings: any): void;
-    /** Should we show the editor? */
-    showEditor?: boolean;
-    /** Useful for playground on squiggle website, where we update the anchor link based on current code and settings */
-    showShareButton?: boolean;
-    /** Height of the editor */
-    height?: number;
-  };
+type PlaygroundProps = // Playground can be either controlled (`code`) or uncontrolled (`defaultCode` + `onCodeChange`)
+  (
+    | { code: string; defaultCode?: undefined }
+    | { defaultCode?: string; code?: undefined }
+  ) &
+    (
+      | {
+          project: SqProject;
+          continues?: string[];
+        }
+      | {}
```

Comment:
seems kind of intense, kudos for getting it to work.

Reply:
This is a simplified version of `useSquiggle` args. I don't like it either, but it's the easiest way to represent props that we actually want on this component, so I don't think it's possible to simplify. Maybe could be made more clear with named types.

In theory, I should've been able to use `Omit<SquiggleArgs, "code">`, but Typescript wasn't smart enough to do Omit on unions.